LLama/Native/Load/NativeLibraryConfig.cs

```diff
@@ -0,0 +1,605 @@
+        /// <param name="Path"></param>
+        /// <param name="Library"></param>
+        /// <param name="UseCuda"></param>
+        /// <param name="AvxLevel"></param>
+        /// <param name="AllowFallback"></param>
+        /// <param name="SkipCheck"></param>
+        /// <param name="SearchDirectories"></param>
+        public record Description(string? Path, NativeLibraryName Library, bool UseCuda, AvxLevel AvxLevel, bool AllowFallback, bool SkipCheck, 
+            string[] SearchDirectories)
+        {
+            /// <inheritdoc/>
+            public override string ToString()
+            {
+                string avxLevelString = AvxLevel switch
+                {
+                    AvxLevel.None => "NoAVX",
+                    AvxLevel.Avx => "AVX",
+                    AvxLevel.Avx2 => "AVX2",
+                    AvxLevel.Avx512 => "AVX512",
+                    _ => "Unknown"
+                };
+
+                string searchDirectoriesString = "{ " + string.Join(", ", SearchDirectories) + " }";
+
+                return $"NativeLibraryConfig Description:\n" +
+                       $"- LibraryName: {Library}\n" +
+                       $"- Path: '{Path}'\n" +
+                       $"- PreferCuda: {UseCuda}\n" +
+                       $"- PreferredAvxLevel: {avxLevelString}\n" +
+                       $"- AllowFallback: {AllowFallback}\n" +
+                       $"- SkipCheck: {SkipCheck}\n" +
+                       $"- SearchDirectories and Priorities: {searchDirectoriesString}";
+            }
+        }
+    }
+#endif
+
+    public sealed partial class NativeLibraryConfig
+    {
+        /// <summary>
+        /// Set configurations for all the native libraries, including LLama and LLava
+        /// </summary>
+        [Obsolete("Please use NativeLibraryConfig.All instead, or set configurations for NativeLibraryConfig.LLama and NativeLibraryConfig.LLavaShared respectively.")]
+        public static NativeLibraryConfigContainer Instance { get; }
+
+        /// <summary>
+        /// Set configurations for all the native libraries, including LLama and LLava
+        /// </summary>
+        public static NativeLibraryConfigContainer All { get; }
+
+        /// <summary>
+        /// Configuration for LLama native library
+        /// </summary>
+        public static NativeLibraryConfig LLama { get; }
+
+        /// <summary>
+        /// Configuration for LLava native library
+        /// </summary>
+        public static NativeLibraryConfig LLavaShared { get; }
+
+        /// <summary>
+        /// A dictionary mapping from version to corresponding llama.cpp commit hash.
+        /// The version should be formatted int `[major].[minor].[patch]`. But there's an exceptance that you can 
+        /// use `master` as a version to get the llama.cpp commit hash from the master branch.
+        /// </summary>
+        public static Dictionary<string, string> VersionMap { get; } = new Dictionary<string, string>() 
```

Comment:
Does this need to be a dictionary? Could it just be a constant, with the commit hash (either of llama.cpp, or llamasharp, or both) hardcoded?

Reply:
Using a dictionary here makes it possible for users to specify another version of native library when auto-download is enabled. This will make the behavior the same with our backend packages -- users could install a backend with version different with the main package.

I used to consider two alternatives for this:
- Hardcode the default version but allows to load a version map from a file outside.
- Just move the version map to #692

Reply:
That's not something that would ever work though? The version of llamasharp is always tied to exactly one version of llama.cpp.

Reply:
What about adding a method to get the llama.cpp commit hash but hide its implementation?

```cs
public string NativeLibraryCommitHash {get;}

private const string version;
private const string commitHash;
```

Thus it will not impact on the public API if we really want to support what I said above in the future.